In [1]:
import json
#import konlpy
import pandas as pd
import numpy as np

#import elice_challenge as ec

In [2]:
train_data = json.load(open('../elice/data_r3/train_data.json'))
test_data = json.load(open('../elice/data_r3/test.input.json'))
grading_data = json.load(open('../elice/data_r3/grading.input.json'))

In [3]:
print(len(train_data))
print(len(test_data))
print(len(grading_data))

76500
4500
9000


In [34]:
total_list, only_text, only_y = [], [], []

#int(len(train_data)/10)
for i in range(int(len(train_data)/5)):
    total_list.append(train_data[i]['datetime'])
    total_list.append(train_data[i]['text'])
    total_list.append(train_data[i]['title'])
    total_list.append(train_data[i]['writing'])
    only_text.append(train_data[i]['text'])
    only_y.append(train_data[i]['writing'])

In [5]:
total_list = np.reshape(total_list, (int(len(train_data)/5), 4))
total_list

array([['20161212',
        '[앵커]청와대 민정수석을 지냈던 우병우씨의 소재를 놓고 설왕설래가 한창입니다. 지난주 청문회 때 출석요구서 수령을 거부해 나오지 않았지요. 국회는 오는 19일 5차 청문회에 우씨를 다시 부르기로 했지만, 그의 소재가 파악되지 않는 한, 불가능한 일입니다. 일각에선 그에게 현상금까지 거는 상황까지 벌어지고 있는데요. 오늘(12일) 국회 발제에선 이 문제를 놓고 얘기해보도록 하겠습니다.   ]네, 그렇습니다. 지난주 \'최순실 국정농단 청문회\'에 우병우씨는 모습을 드러내지 않았습니다. 다른 사람들처럼 출석요구서를 전달받고 불출석사유를 낸 것도 아니고, 아예 잠적을 해버려서, 요구서 전달조차 안 된 상황이었던 겁니다. 어떻게 보면 질이 가장 안 좋은 상황이라고 볼 수 있는 거죠.보다 못한 몇몇분들이 "우병우를 잡겠다"고 하면서 현상금을 걸었습니다.정봉주 전 의원, 안민석 의원이 각각 500만원, 김성태 \'최순실 국정조사 특위위원장\'이 100만원, 또 주진우 가 200만원, 이렇게 도합 1300만원이 걸린 겁니다.심지어는 많은 시민들이 \'나도 동참하고 싶다\'면서 계좌를 열어라! 하셨다는군요. 그래서 일종의 \'펀딩\' 형식으로 액수가 계속 불어날 것 같다고 합니다.다른 분들은 그런 줄 알겠는데, 안민석 의원이 500만원을 걸었다, 이건 좀 충격적인데요. 저번에 소개해드렸지만, 장시호 잡겠다고 저랑 같이 야밤에 여기저기 돌아다닐 때, 배가 너무 고파서, 같이 24시간 하는 식당 가서 떡만둣국을 먹었거든요. 잘 먹고 나서 계산하려고 하니까, 갑자기 "지갑을 차에 두고 왔다"고 하는 겁니다.이날 밥값을 제가 냈습니다. 아직 영수증도 갖고 있습니다. 그러셨던 분이 500만원을 걸었답니다.자, 아무튼 제가 현상금을 맨 처음 제안한 정봉주 전 의원에게 전화해봤습니다. 뭔가 그럴듯한 제보가 들어왔는지 말이죠. 그랬더니 이렇게 말합니다.[정봉주/전 의원 : 종로구 옥인동 소재에 우병우 닮은 사람이 마당에서 서성인다는 걸 

In [6]:
train_df = pd.DataFrame(total_list, columns=['datetime', 'text', 'title', 'writing'])
train_df.tail()

datetime                                               text  \
15295  20160506  여야 3당의 원내대표가 확정되면서 20대 국회 상임위원회 쟁탈전이 가시화하고 있다....   
15296  20160908  "사법개혁특별위 설치 의장 주재 3당 원내대표 회동해야"    (서울=<언론사명>)...   
15297  20160718  <앵커 멘트> 몽골을 방문 중인 박근혜 대통령은 마지막 일정으로  몽골 국회의장과 ...   
15298  20160929  사드 제3후보지 발표 임박…군, 후폭풍 고심[앵커] 사드 제3후보지 발표가 임박한 ...   
15299  20160319  [<언론사명>]내년부터 일본의 고교 1학년생이 배우게 될 사회과 교과서 35종 중 ...   

                                title writing  
15295       與野, 쟁점법안 `마지막 관문` 법사위 쟁탈전    매일경제  
15296  박지원 "朴대통령, 추석후 3당대표 회동 약속 지켜야"    연합뉴스  
15297                           동영상기사  KBS 뉴스  
15298                           동영상기사  연합뉴스TV  
15299   [사설]독도·위안부 왜곡 교과서로 日미래세대 키워서야    동아일보

In [28]:
# split data into X and y
#dfX = train_df.iloc[:,:-1]
dfX = train_df.iloc[:,1:2]
dfy = train_df.iloc[:,[-1]]

In [7]:
train_df2 = train_df.groupby('writing')
train_df2.count()

datetime   text  title
writing                        
JTBC         2416   2416   2416
KBS 뉴스       6285   6285   6285
MBN          7399   7399   7399
TV조선         1359   1359   1359
YTN          4089   4089   4089
경향신문         1535   1535   1535
국민일보         5594   5594   5594
노컷뉴스          471    471    471
뉴시스         12363  12363  12363
동아일보         4981   4981   4981
매일경제         3135   3135   3135
서울신문         4057   4057   4057
아시아경제         559    559    559
연합뉴스         7863   7863   7863
연합뉴스TV       3408   3408   3408
오마이뉴스        1934   1934   1934
채널A          2218   2218   2218
파이낸셜뉴스       4181   4181   4181
헤럴드경제        2653   2653   2653

### 언론사 수
- 언론사 19가지

### 언론사별 데이터 수
- JTBC : 2416
- KBS 뉴스 : 6285
- MBN : 7399
- TV조선 : 1359
- YTN : 4089
- 경향신문 : 1535
- 국민일보 : 5594
- 노컷뉴스 : 471
- 뉴시스 : 12363
- 동아일보 : 4981
- 매일경제 : 3135
- 서울신문 : 4057
- 아시아경제 : 559
- 연합뉴스 : 7863
- 연합뉴스TV : 3408
- 오마이뉴스 : 1934
- 채널A : 2218
- 파이낸셜뉴스 : 4181
- 헤럴드경제 : 2653

In [8]:
train_data[0]

{'datetime': '20161212',
 'text': '[앵커]청와대 민정수석을 지냈던 우병우씨의 소재를 놓고 설왕설래가 한창입니다. 지난주 청문회 때 출석요구서 수령을 거부해 나오지 않았지요. 국회는 오는 19일 5차 청문회에 우씨를 다시 부르기로 했지만, 그의 소재가 파악되지 않는 한, 불가능한 일입니다. 일각에선 그에게 현상금까지 거는 상황까지 벌어지고 있는데요. 오늘(12일) 국회 발제에선 이 문제를 놓고 얘기해보도록 하겠습니다.   ]네, 그렇습니다. 지난주 \'최순실 국정농단 청문회\'에 우병우씨는 모습을 드러내지 않았습니다. 다른 사람들처럼 출석요구서를 전달받고 불출석사유를 낸 것도 아니고, 아예 잠적을 해버려서, 요구서 전달조차 안 된 상황이었던 겁니다. 어떻게 보면 질이 가장 안 좋은 상황이라고 볼 수 있는 거죠.보다 못한 몇몇분들이 "우병우를 잡겠다"고 하면서 현상금을 걸었습니다.정봉주 전 의원, 안민석 의원이 각각 500만원, 김성태 \'최순실 국정조사 특위위원장\'이 100만원, 또 주진우 가 200만원, 이렇게 도합 1300만원이 걸린 겁니다.심지어는 많은 시민들이 \'나도 동참하고 싶다\'면서 계좌를 열어라! 하셨다는군요. 그래서 일종의 \'펀딩\' 형식으로 액수가 계속 불어날 것 같다고 합니다.다른 분들은 그런 줄 알겠는데, 안민석 의원이 500만원을 걸었다, 이건 좀 충격적인데요. 저번에 소개해드렸지만, 장시호 잡겠다고 저랑 같이 야밤에 여기저기 돌아다닐 때, 배가 너무 고파서, 같이 24시간 하는 식당 가서 떡만둣국을 먹었거든요. 잘 먹고 나서 계산하려고 하니까, 갑자기 "지갑을 차에 두고 왔다"고 하는 겁니다.이날 밥값을 제가 냈습니다. 아직 영수증도 갖고 있습니다. 그러셨던 분이 500만원을 걸었답니다.자, 아무튼 제가 현상금을 맨 처음 제안한 정봉주 전 의원에게 전화해봤습니다. 뭔가 그럴듯한 제보가 들어왔는지 말이죠. 그랬더니 이렇게 말합니다.[정봉주/전 의원 : 종로구 옥인동 소재에 우병우 닮은 사람이 마당에서 서성

In [31]:
from sklearn.feature_extraction.text import CountVectorizer
vect = CountVectorizer()
vect.fit(only_text)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)

In [32]:
bag_of_words = vect.transform(only_text)
print("bag_of_words:\n{}".format(repr(bag_of_words)))

bag_of_words:
<15300x358791 sparse matrix of type '<class 'numpy.int64'>'
	with 2615980 stored elements in Compressed Sparse Row format>


In [36]:
feature_names = vect.get_feature_names()
print('특성 개수: {}'.format(len(feature_names)))
print('처음 20개 특성: {}'.format(feature_names[:20]))
print('10000부터 10020까지 특성: {}'.format(feature_names[10000:10020]))

특성 개수: 358791
처음 20개 특성: ['00', '000', '00016', '00018', '0003', '0003w', '0007', '0007w', '000km', '000km가', '000km까지', '000km를', '000km에', '000km의', '000m', '000mw', '000m에서', '000t급', '000t급의', '000t의']
10000부터 10020까지 특성: ['2시36분', '2시40분', '2시40분께', '2시45분까지', '2시45분부터', '2시45분이었는데요', '2시50분께', '2시50분여', '2시56분께', '2시가', '2시간', '2시간10분', '2시간15분', '2시간15분이나', '2시간20분', '2시간20분한미', '2시간30분', '2시간45분만이었다', '2시간가량', '2시간까지']


In [35]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression

scores = cross_val_score(LogisticRegression(), bag_of_words, only_y, cv=5 )
print('교차 검증 평균 점수 : {:.2f}'.format(np.mean(scores)))

교차 검증 평균 점수 : 0.66
